### DISCUSSION TITLE
Explainability problem

### DISCUSSION THREAD

**Discussion by Rabea007** [timeseriesAI/tsai#559] [Q&A]
Dear tsai team,
We are facing a problem in the explainability section of XCM model, from this tutorial:
https://colab.research.google.com/github/timeseriesAI/tsai/blob/master/nbs/114_models.XCM.ipynb
We get 2 difference attribution maps when we run the show_gradcam function.

Using the same X, one time we provided the y, and the second time we didn't (y=None), and we got 2 different heatmaps: 
`model.show_gradcam(X,y)`

y=None:  
<img width="891" alt="Screen Shot 2022-08-03 at 16 33 03" src="https://user-images.githubusercontent.com/93527567/182779080-d19ce918-6fae-48df-8b6c-64dd126eddaf.png">

y=0: 
<img width="911" alt="Screen Shot 2022-08-03 at 16 32 54" src="https://user-images.githubusercontent.com/93527567/182779109-b4d63c6f-d647-4d8f-840e-2373c635442d.png">

digging deep inside the function, in order to understand this behavior. We found that inside get_acts_and_grads  you guys are using this line:
`preds = model.eval()(x)`

which doesn't really correspond to what we get from the get_X_preds() function and this might be what is making the 
difference.

 _______________________________________________
```
def get_acts_and_grads(model, modules, x, y=None, detach=True, cpu=False):
    r"""Returns activations and gradients for given modules in a model and a single input or a batch.
    Gradients require y value(s). If they rae not provided, it will use the predicttions. """
    if not is_listy(modules): modules = [modules]
    x = x[None, None] if x.ndim == 1 else x[None] if x.ndim == 2 else x
    with hook_outputs(modules, detach=detach, cpu=cpu) as h_act:
        with hook_outputs(modules, grad=True, detach=detach, cpu=cpu) as h_grad:
            preds = model.eval()(x)
            if y is None: preds.max(dim=-1).values.mean().backward()
            else:
                if preds.shape[0] == 1: preds[0, y].backward()
                else:
                    if y.ndim == 1: y = y.reshape(-1, 1)
                    torch_slice_by_dim(preds, y).mean().backward()
    if len(modules) == 1: return h_act.stored[0].data, h_grad.stored[0][0].data
    else: return [h.data for h in h_act.stored], [h[0].data for h in h_grad.stored]

```
  
for example:
```
x = TSTensor(samples:1, vars:10, len:128, device=cpu, dtype=torch.float32) 
y = TensorCategory(0) ( got this y using learn.get_X_preds(X, with_decoded=True) )
```

Calling the get_acts_and_grads function the first time ( y is provided, and not None):
running` preds = model.eval()(x)` give us :
```
preds = TensorBase([[-2.1717,  1.9500]], grad_fn=<AliasBackward0>)
hence, preds.max(dim=-1) results in: TensorBase([1])
```

which is different from the original y!! Since y is provided, y is not None and we we will backward on preds[0, y] which equals to -2.1717
When y is not provided (y is None), we will backward on "preds.max(dim=-1).values.mean()" which in this case equals 1.9500. 

p.s. after training the model, these 2 lines yield different results ( y_pred1 doesn't equal y_pred2)
(1) ```preds = model.eval()(X)
        y_pred1=preds.max(dim=-1)```
    (2) `_,_, y_pred2 = learn.get_X_preds(X, with_decoded=True) ` 

So this might be the bug, because these 2 outputs should not be different, because seems like this is what you guys assumed when this function was built ) 
I would really appreciate your response, because we are stuck on this bug for a while now, 
Many thanks,

**Comment by oguiza** [comment]
There was a bug in the learn.get_X_preds code that has been fixed. I believe the issue you were facing should also be fixed.
